ReactiveMessagePipelineTest.handlingTimeout is flaky

https://github.com/apache/pulsar-client-reactive/actions/runs/3641666977/jobs/6147871691#step:4:565

```
  Test handlingTimeout() FAILED

  java.lang.AssertionError: 
  Expected size: 9 but was: 8 in:
  [123456:4128:-1,
      123456:4129:-1,
      123456:4130:-1,
      123456:4131:-1,
      123456:4133:-1,
      123456:4134:-1,
      123456:4135:-1,
      123456:4136:-1]
      at org.apache.pulsar.reactive.client.api.ReactiveMessagePipelineTest.handlingTimeout(ReactiveMessagePipelineTest.java:208)


    [2022-12-07 18:20:31,485] [parallel-2] [org.apache.pulsar.reactive.client.internal.api.DefaultReactiveMessagePipeline] ERROR Message handling for message id 123456:4127:-1 failed.
    java.util.concurrent.TimeoutException: Did not observe any item or terminal signal within 5ms in 'source(MonoDefer)' (and no fallback has been configured)
    	at reactor.core.publisher.FluxTimeout$TimeoutMainSubscriber.handleTimeout(FluxTimeout.java:295) ~[reactor-core-3.4.22.jar:3.4.22]
    	at reactor.core.publisher.FluxTimeout$TimeoutMainSubscriber.doTimeout(FluxTimeout.java:280) ~[reactor-core-3.4.22.jar:3.4.22]
    	at reactor.core.publisher.FluxTimeout$TimeoutTimeoutSubscriber.onNext(FluxTimeout.java:419) ~[reactor-core-3.4.22.jar:3.4.22]
    	at reactor.core.publisher.FluxOnErrorResume$ResumeSubscriber.onNext(FluxOnErrorResume.java:79) ~[reactor-core-3.4.22.jar:3.4.22]
    	at reactor.core.publisher.MonoDelay$MonoDelayRunnable.propagateDelay(MonoDelay.java:271) ~[reactor-core-3.4.22.jar:3.4.22]
    	at reactor.core.publisher.MonoDelay$MonoDelayRunnable.run(MonoDelay.java:286) ~[reactor-core-3.4.22.jar:3.4.22]
    	at reactor.core.scheduler.SchedulerTask.call(SchedulerTask.java:68) ~[reactor-core-3.4.22.jar:3.4.22]
    	at reactor.core.scheduler.SchedulerTask.call(SchedulerTask.java:28) ~[reactor-core-3.4.22.jar:3.4.22]
    	at java.util.concurrent.FutureTask.run(FutureTask.java:264) ~[?:?]
    	at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:304) ~[?:?]
    	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128) ~[?:?]
    	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628) ~[?:?]
    	at java.lang.Thread.run(Thread.java:829) ~[?:?]
    [2022-12-07 18:20:31,500] [parallel-2] [org.apache.pulsar.reactive.client.internal.api.DefaultReactiveMessagePipeline] ERROR Message handling for message id 123456:4132:-1 failed.
    java.util.concurrent.TimeoutException: Did not observe any item or terminal signal within 5ms in 'source(MonoDefer)' (and no fallback has been configured)
    	at reactor.core.publisher.FluxTimeout$TimeoutMainSubscriber.handleTimeout(FluxTimeout.java:295) ~[reactor-core-3.4.22.jar:3.4.22]
    	at reactor.core.publisher.FluxTimeout$TimeoutMainSubscriber.doTimeout(FluxTimeout.java:280) ~[reactor-core-3.4.22.jar:3.4.22]
    	at reactor.core.publisher.FluxTimeout$TimeoutTimeoutSubscriber.onNext(FluxTimeout.java:419) ~[reactor-core-3.4.22.jar:3.4.22]
    	at reactor.core.publisher.FluxOnErrorResume$ResumeSubscriber.onNext(FluxOnErrorResume.java:79) ~[reactor-core-3.4.22.jar:3.4.22]
    	at reactor.core.publisher.MonoDelay$MonoDelayRunnable.propagateDelay(MonoDelay.java:271) ~[reactor-core-3.4.22.jar:3.4.22]
    	at reactor.core.publisher.MonoDelay$MonoDelayRunnable.run(MonoDelay.java:286) ~[reactor-core-3.4.22.jar:3.4.22]
    	at reactor.core.scheduler.SchedulerTask.call(SchedulerTask.java:68) ~[reactor-core-3.4.22.jar:3.4.22]
    	at reactor.core.scheduler.SchedulerTask.call(SchedulerTask.java:28) ~[reactor-core-3.4.22.jar:3.4.22]
    	at java.util.concurrent.FutureTask.run(FutureTask.java:264) ~[?:?]
    	at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:304) ~[?:?]
    	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128) ~[?:?]
    	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628) ~[?:?]
    	at java.lang.Thread.run(Thread.java:829) ~[?:?]
    [2022-12-07 18:20:31,509] [parallel-1] [org.apache.pulsar.reactive.client.internal.api.DefaultReactiveMessagePipeline] ERROR ReactiveMessageHandler was unexpectedly completed.

```

--- a/pulsar-client-reactive-api/src/test/java/org/apache/pulsar/reactive/client/api/ReactiveMessagePipelineTest.java
+++ b/pulsar-client-reactive-api/src/test/java/org/apache/pulsar/reactive/client/api/ReactiveMessagePipelineTest.java
@@ -186,7 +186,8 @@ class ReactiveMessagePipelineTest {
 	void handlingTimeout() throws Exception {
 		int numMessages = 10;
 		TestConsumer testConsumer = new TestConsumer(numMessages);
-		CountDownLatch latch = new CountDownLatch(numMessages);
+		CountDownLatch latch = new CountDownLatch(1);
+		testConsumer.setFinishedCallback(latch::countDown);
 		AtomicReference<MessageId> timedoutMessageId = new AtomicReference<>();
 		Function<Message<String>, Publisher<Void>> messageHandler = (message) -> Mono.defer(() -> {
 			Duration delay;
@@ -197,13 +198,12 @@ class ReactiveMessagePipelineTest {
 			else {
 				delay = Duration.ofMillis(2);
 			}
-			return Mono.delay(delay).doFinally((__) -> latch.countDown()).then();
+			return Mono.delay(delay).then();
 		});
 		try (ReactiveMessagePipeline pipeline = testConsumer.messagePipeline().messageHandler(messageHandler)
 				.handlingTimeout(Duration.ofMillis(5)).build()) {
 			pipeline.start();
 			assertThat(latch.await(1, TimeUnit.SECONDS)).isTrue();
-			pipeline.stop();
 			// 9 messages should have been acked
 			assertThat(testConsumer.getAcknowledgedMessages()).hasSize(9);
 			// 1 message should have been nacked
@@ -422,6 +422,8 @@ class ReactiveMessagePipelineTest {
 
 		private final int numMessages;
 
+		private volatile Runnable finishedCallback;
+
 		TestConsumer(int numMessages) {
 			this.numMessages = numMessages;
 		}
@@ -429,6 +431,10 @@ class ReactiveMessagePipelineTest {
 		private final List<MessageId> acknowledgedMessages = new CopyOnWriteArrayList<>();
 
 		private final List<MessageId> nackedMessages = new CopyOnWriteArrayList<>();
+
+		void setFinishedCallback(Runnable finishedCallback) {
+			this.finishedCallback = finishedCallback;
+		}
 
 		@Override
 		public <R> Flux<R> consumeMany(Function<Flux<Message<String>>, Publisher<MessageResult<R>>> messageHandler) {
@@ -442,7 +448,11 @@ class ReactiveMessagePipelineTest {
 					else {
 						this.nackedMessages.add(result.getMessageId());
 					}
-				}).mapNotNull(MessageResult::getValue);
+				}).mapNotNull(MessageResult::getValue).doFinally((__) -> {
+					if (this.finishedCallback != null) {
+						this.finishedCallback.run();
+					}
+				});
 			});
 		}
 

